kernel/printf.c

```diff
@@ -67,9 +63,9 @@ printf(char *fmt, ...)
   int i, c, locking;
   char *s;
 
-  locking = pr.locking;
-  if(locking)
-    acquire(&pr.lock);
+  locking = cons.locking;
+  if(locking && !holding(&cons.lock))
```

Comment:
Two different threads can acquire lock here. I think this is why you separated cons.lock and pr.lock. I'll close pull request.